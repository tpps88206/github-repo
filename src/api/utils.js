export const getHeaders = () => ({
  Accept: 'application/vnd.github.v3+json',
  'Content-Type': 'Application/json',
});
