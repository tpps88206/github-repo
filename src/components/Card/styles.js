const styles = theme => ({
  root: {
    margin: 12,
    height: 284,
  },
  '@media screen and (min-width: 1280px)': {
    root: {
      maxWidth: 600,
      minWidth: 400,
    },
  },
  '@media screen and (min-width: 480px) and (max-width: 1279px)': {
    root: {
      width: 400,
    },
  },
  '@media screen and (max-width: 479px)': {
    root: {
      width: 'calc(100vw - 24px)',
    },
  },
  avatar: {
    color: 'white',
    backgroundColor: theme.palette.primary.main,
  },
  description: {
    overflowY: 'auto',
    height: 44,
  },
  content: {
    height: 155,
  },
  starts: {
    lineHeight: '40px',
  },
  forks: {
    lineHeight: '40px',
  },
  language: {
    lineHeight: '40px',
  },
});

export default styles;
