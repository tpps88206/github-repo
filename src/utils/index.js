import isEmpty from 'lodash/isEmpty';
import split from 'lodash/split';
import upperCase from 'lodash/upperCase';

export const generateAvatarCharacter = fullName => {
  if (!fullName) {
    return '';
  }
  const repoName = split(fullName, '/', 2);
  if (repoName.length > 1) {
    return upperCase(repoName[1].substring(0, 1));
  }
  return upperCase(repoName[0].substring(0, 1));
};

export const generateQueryParams = params => {
  let result = '';
  if (isEmpty(params)) {
    return result;
  }

  params.forEach((param, index) => {
    if (index > 0) {
      result = result + '&';
    }
    result = result + param.key + '=' + param.value;
  });
  return result;
};
