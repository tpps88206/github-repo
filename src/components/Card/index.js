import React, { memo, useState } from 'react';
import { useDispatch } from 'react-redux';

import has from 'lodash/has';
import upperCase from 'lodash/upperCase';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Divider } from '@material-ui/core';
import Avatar from '@material-ui/core/Avatar';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import FavoriteIcon from '@material-ui/icons/Favorite';
import FavoriteBorderIcon from '@material-ui/icons/FavoriteBorder';
import ShareIcon from '@material-ui/icons/Share';

import { ICONS } from '@/constants/icons';
import { SUCCESS_TYPE } from '@/constants/variables';
import { addNotification } from '@/redux/slices/notification';
import styles from './styles';

const useStyles = makeStyles(styles);

const EnhancedCard = ({ fullName, description, updatedAt, htmlUrl, homepage, language, forks, watchers, ...props }) => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const [favorite, setFavorite] = useState(false);

  const handleClickFavoriteButton = () => {
    setFavorite(favorite => !favorite);
  };

  const handleClickShareButton = () => {
    const element = document.createElement('textarea');
    element.value = htmlUrl;
    document.body.appendChild(element);
    element.select();
    document.execCommand('copy');
    document.body.removeChild(element);

    dispatch(
      addNotification({
        notification: {
          severity: SUCCESS_TYPE,
          message: '已將網址複製到剪貼簿上',
        },
      }),
    );
  };

  const renderLanguageSection = () => {
    if (has(ICONS, language)) {
      return (
        <React.Fragment>
          <Typography variant="body2" color="textSecondary" component="p">
            Language
          </Typography>
          <FontAwesomeIcon icon={ICONS[language]} size="3x" />
        </React.Fragment>
      );
    } else {
      return (
        <React.Fragment>
          <Typography variant="body2" color="textSecondary" component="p">
            Language
          </Typography>
          <Typography variant="subtitle1" color="primary" component="p" className={classes.language}>
            {language ? language : '未知'}
          </Typography>
        </React.Fragment>
      );
    }
  };

  return (
    <Grid item lg={4}>
      <Card className={classes.root} {...props}>
        <CardHeader
          avatar={
            <Avatar aria-label="recipe" className={classes.avatar}>
              {fullName && upperCase(fullName.substring(0, 1))}
            </Avatar>
          }
          title={fullName}
          subheader={'最後更新：' + updatedAt}
        />
        <CardContent className={classes.content}>
          <div className={classes.description}>
            <Typography variant="body2" color="textPrimary" component="p">
              {description}
            </Typography>
          </div>
          <Divider className="my-1" />
          <div>
            <Grid container spacing={1}>
              <Grid item xs={4} className="text-center">
                {renderLanguageSection()}
              </Grid>
              <Grid item xs={4} className="text-center">
                <Typography variant="body2" color="textSecondary" component="p">
                  Stars
                </Typography>
                <Typography variant="subtitle1" color="primary" component="p" className={classes.starts}>
                  {watchers}
                </Typography>
              </Grid>
              <Grid item xs={4} className="text-center">
                <Typography variant="body2" color="textSecondary" component="p">
                  Forks
                </Typography>
                <Typography variant="subtitle1" color="primary" component="p" className={classes.forks}>
                  {forks}
                </Typography>
              </Grid>
            </Grid>
          </div>
        </CardContent>
        <CardActions disableSpacing>
          {favorite ? (
            <IconButton aria-label="remove from favorites" color="secondary" onClick={handleClickFavoriteButton}>
              <FavoriteIcon />
            </IconButton>
          ) : (
            <IconButton aria-label="add to favorites" color="secondary" onClick={handleClickFavoriteButton}>
              <FavoriteBorderIcon />
            </IconButton>
          )}
          <IconButton aria-label="share" color="secondary" onClick={handleClickShareButton}>
            <ShareIcon />
          </IconButton>
        </CardActions>
      </Card>
    </Grid>
  );
};

export default memo(EnhancedCard);
