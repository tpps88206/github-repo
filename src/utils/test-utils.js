import React from 'react';
import { Provider } from 'react-redux';
// if you use Jest for testing and get the error with "use <Link> outside a <Router>"
// because just wrap your component with <BrowserRouter>
// see https://stackoverflow.com/questions/48640280/you-should-not-use-link-outside-a-router
import { BrowserRouter } from 'react-router-dom';

import { create } from 'jss';
import preset from 'jss-preset-default';

import { library } from '@fortawesome/fontawesome-svg-core';
import { fab } from '@fortawesome/free-brands-svg-icons';
import CssBaseline from '@material-ui/core/CssBaseline';
import { createMuiTheme, StylesProvider, ThemeProvider } from '@material-ui/core/styles';
import { render as rtlRender } from '@testing-library/react';

import theme from '@/constants/theme';
import configureStore from '@/redux/configureStore';
import styles from '@/styles';

library.add(fab);

const render = (ui, { ...renderOptions } = {}) => {
  const muiTheme = createMuiTheme(theme);
  const jss = create({
    plugins: [...preset().plugins],
  });
  const store = configureStore();

  jss.createStyleSheet(styles(muiTheme), { meta: 'global' }).attach();

  const Wrapper = ({ children }) => {
    return (
      <StylesProvider jss={jss}>
        <ThemeProvider theme={muiTheme}>
          <CssBaseline />
          <Provider store={store}>
            <BrowserRouter>{children}</BrowserRouter>
          </Provider>
        </ThemeProvider>
      </StylesProvider>
    );
  };
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
