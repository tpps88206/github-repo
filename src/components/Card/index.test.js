import React from 'react';

import { render } from '@/utils/test-utils';
import Card from './index';

describe('test Card', () => {
  it('renders Card without crashing', () => {
    const div = document.createElement('div');
    render(<Card />, div);
  });

  it('renders Card with snapshot', () => {
    const { asFragment } = render(<Card />);
    expect(asFragment()).toMatchSnapshot();
  });
});
