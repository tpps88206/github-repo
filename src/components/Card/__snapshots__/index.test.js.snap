// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Card renders Card with snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-4"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-8 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar"
        >
          <div
            aria-label="recipe"
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-9 MuiAvatar-colorDefault"
          />
        </div>
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            最後更新：undefined
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root makeStyles-content-11"
      >
        <div
          class="makeStyles-description-10"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
          />
        </div>
        <hr
          class="MuiDivider-root my-1"
        />
        <div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
          >
            <div
              class="MuiGrid-root text-center MuiGrid-item MuiGrid-grid-xs-4"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                Language
              </p>
              <p
                class="MuiTypography-root makeStyles-language-14 MuiTypography-subtitle1 MuiTypography-colorPrimary"
              >
                未知
              </p>
            </div>
            <div
              class="MuiGrid-root text-center MuiGrid-item MuiGrid-grid-xs-4"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                Stars
              </p>
              <p
                class="MuiTypography-root makeStyles-starts-12 MuiTypography-subtitle1 MuiTypography-colorPrimary"
              />
            </div>
            <div
              class="MuiGrid-root text-center MuiGrid-item MuiGrid-grid-xs-4"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                Forks
              </p>
              <p
                class="MuiTypography-root makeStyles-forks-13 MuiTypography-subtitle1 MuiTypography-colorPrimary"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiCardActions-root"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="share"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
